src/include/txn.i

```diff
@@ -768,15 +781,18 @@ static inline int
 __wt_txn_read_upd_list(WT_SESSION_IMPL *session, WT_UPDATE *upd, WT_UPDATE **updp)
 {
     WT_VISIBLE_TYPE upd_visible;
+    uint8_t type;
 
     *updp = NULL;
+
     for (; upd != NULL; upd = upd->next) {
+        WT_ORDERED_READ(type, upd->type);
         /* Skip reserved place-holders, they're never visible. */
-        if (upd->type == WT_UPDATE_RESERVE)
+        if (type == WT_UPDATE_RESERVE)
             continue;
         upd_visible = __wt_txn_upd_visible_type(session, upd);
         if (upd_visible == WT_VISIBLE_TRUE) {
-            if (upd->type != WT_UPDATE_BIRTHMARK)
+            if (type != WT_UPDATE_BIRTHMARK)
```

Comment:
This looks correct to me (and is still needed, the check at line 795 can hit the race).

I think the specific line you asked about:
```
WT_ASSERT(session, upd == NULL || upd->type != WT_UPDATE_BIRTHMARK);
```
is safe because it's only checking an update returned from `__wt_find_hs_upd()` as far as I can tell, which is returning an allocated `WT_UPDATE` structure.

FWIW, I think the usage of the local `upd` variable in `__wt_txn_read()` is confusing at this point and could be removed. Here's what I'm suggesting:
```
*** src/include/txn.i   Tue Feb  4 23:22:54 2020
--- -   Tue Feb  4 23:32:26 2020 
***************  
*** 823,830 **** 
      if (*updp != NULL)
          return (0);
          
-     upd = NULL;
-     
      /* If there is no ondisk value, there can't be anything in the history store either. */
      if (cbt->ref->page->dsk == NULL || cbt->slot == UINT32_MAX) {
          WT_RET(__upd_alloc_tombstone(session, updp, WT_TXN_NONE, WT_TS_NONE));
--- 823,828 ----
*************** 
*** 853,873 ****
       * the current system.
       */
      if (__wt_txn_visible(session, start.txnid, start.timestamp)) {
!         WT_RET(__wt_update_alloc(session, &buf, &upd, &size, WT_UPDATE_STANDARD));
!         upd->txnid = start.txnid;
!         upd->start_ts = upd->durable_ts = start.timestamp;
!         F_SET(upd, WT_UPDATE_RESTORED_FROM_DISK);
!         *updp = upd;
          return (0); 
      }   
      
      /* If there's no visible update in the update chain or ondisk, check the history store file. *
/     
!     WT_ASSERT(session, upd == NULL);
!     if (F_ISSET(S2C(session), WT_CONN_HS_OPEN) && !F_ISSET(S2BT(session), WT_BTREE_HS))
!         WT_RET_NOTFOUND_OK(__wt_find_hs_upd(session, cbt, &upd, false, &buf, &start));

!     /* There is no BIRTHMARK in the history store file. */
!     WT_ASSERT(session, upd == NULL || upd->type != WT_UPDATE_BIRTHMARK);

      /*
       * If we checked the update list, the ondisk value and the history store, we should return a
--- 851,870 ----
       * the current system.
       */
      if (__wt_txn_visible(session, start.txnid, start.timestamp)) {
!         WT_RET(__wt_update_alloc(session, &buf, updp, &size, WT_UPDATE_STANDARD));
!         (*updp)->txnid = start.txnid;
!         (*updp)->start_ts = upd->durable_ts = start.timestamp;
!         F_SET((*updp), WT_UPDATE_RESTORED_FROM_DISK);
          return (0);
      }

      /* If there's no visible update in the update chain or ondisk, check the history store file. *
/
!     if (F_ISSET(S2C(session), WT_CONN_HS_OPEN) && !F_ISSET(S2BT(session), WT_BTREE_HS)) {
!         WT_RET_NOTFOUND_OK(__wt_find_hs_upd(session, cbt, updp, false, &buf, &start));

!       /* There is no BIRTHMARK in the history store file. */
!       WT_ASSERT(session, (*updp)->type != WT_UPDATE_BIRTHMARK);
!     }

      /*
       * If we checked the update list, the ondisk value and the history store, we should return a
***************
*** 876,883 ****
       * FIXME-PM-1521: We call transaction read in a lot of places so we can't do this yet. When we
       * refactor this function to return a byte array, we should tackle this at the same time.
       */
-
-     *updp = upd;
      return (0);
  }

--- 873,878 ----
```